edx_rbac/utils.py

```diff
@@ -0,0 +1,102 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+import crum
+
+from edx_rest_framework_extensions.auth.jwt.cookies import (
+    jwt_cookie_name,
+)
+from edx_rest_framework_extensions.auth.jwt.decoder import jwt_decode_handler
+from edx_rest_framework_extensions.auth.jwt.authentication import JwtAuthentication
+
+from django.conf import settings
+
+
+def get_request_or_stub():
+    """
+    Return the current request or a stub request.
+    If called outside the context of a request, construct a fake
+    request that can be used to build an absolute URI.
+    This is useful in cases where we need to pass in a request object
+    but don't have an active request (for example, in tests, celery tasks, and XBlocks).
+    """
+    request = crum.get_current_request()
+
+    if request is None:
+
+        # The settings SITE_NAME may contain a port number, so we need to
+        # parse the full URL.
+        full_url = "http://{site_name}".format(site_name=settings.SITE_NAME)
+        parsed_url = urlparse(full_url)
+
+        # Construct the fake request.  This can be used to construct absolute
+        # URIs to other paths.
+        return RequestFactory(
+            SERVER_NAME=parsed_url.hostname,
+            SERVER_PORT=parsed_url.port or 80,
+        ).get("/")
+
+    else:
+        return request
+
+
+def get_decoded_jwt_from_request(request):
+    """
+    Grabs jwt from request if possible.
+
+    Returns a decoded jwt dict if it finds it.
+    Returns a None if it does not.
+    """
+    jwt_cookie_name = settings.JWT_AUTH.get('JWT_AUTH_COOKIE') or 'edx-jwt-cookie'
+    jwt_cookie = request.COOKIES.get(jwt_cookie_name, None)
+    
+    if not jwt_cookie:
+        return None
+    return jwt_decode_handler(jwt_cookie)
+
+
+def request_user_has_implicit_access_via_jwt(decoded_jwt, role_name):
+    """
+    Check the request's user access by mapping user's roles found in jwt
+    to local feature roles.
+
+    decoded_jwt is a dict
+    role_name is a string
+
+    Returns a boolean.
+
+    Mapping should be in settings and look like:
+
+        SYSTEM_TO_FEATURE_ROLE_MAPPING = {
+            'enterprise_admin': ['coupon-management', 'data_api_access'],
+            'enterprise_leaner': [],
+            'portal_admin_no_coupons': ['data_api_access']
+        }
+    """
+    jwt_roles_claim = decoded_jwt.get('roles', [])
+
+    feature_roles = []
+    for role_data in jwt_roles_claim:
+        role_in_jwt, object_type, object_key = role_data.split(':') # split should be more robust because of our cousekeys having colons
```

Comment:
i haven't implemented it yet, but i think we'll probably get rid of the object_type and just have role:context_data

Reply:
cool, noted!